src/huggingface_hub/hf_api.py

```diff
@@ -1615,46 +1615,133 @@ def run_as_future(self, fn: Callable[..., R], *args, **kwargs) -> Future[R]:
         self._thread_pool
         return self._thread_pool.submit(fn, *args, **kwargs)
 
+
+    from huggingface_hub.utils._auth import _get_token_from_google_colab, _get_token_from_environment, _get_token_from_file
+
+    def _get_token_source(self, token: Optional[Union[bool, str]] = None) -> str:
+        """
+        Determine the source of the token being used.
+        
+        Args:
+            token: The token parameter passed to whoami
+            
+        Returns:
+            str: Source of the token ("parameter", "environment", "file", "colab", or "unknown")
+        """
+        if isinstance(token, str):
+            return "parameter"
+        
+        # Use the helper functions to check token sources
+        if _get_token_from_environment():
+            return "environment"
+        
+        if _get_token_from_file():
+            return "file"
+        
+        if _get_token_from_google_colab():
+            return "colab"
+        
+        return "unknown"
+
+    def _get_token_error_message(self, source: str) -> str:
+        """
+        Generate a detailed error message based on the token source.
+        
+        Args:
+            source: The source of the token ("parameter", "environment", "file", "colab", or "unknown")
+            
+        Returns:
+            str: Customized error message
+        """
+        base_message = "Invalid user token. "
+        
+        if source == "parameter":
+            return base_message + (
+                "The token you provided as parameter is invalid. Please verify it's correct "
+                "or remove it to use the default authentication method."
+            )
+        elif source == "environment":
+            return base_message + (
+                "The token from HF_TOKEN environment variable is invalid. "
+                "Note that HF_TOKEN takes precedence over the token file. "
+                "Either update HF_TOKEN or unset it to use the token from `huggingface-cli login`."
+            )
+        elif source == "file":
+            return base_message + (
+                "The token stored in ~/.cache/huggingface/token is invalid. "
+                "Please run `huggingface-cli login` to update it."
+            )
+        elif source == "colab":
+            return base_message + (
+                "The token from Google Colab vault is invalid. "
+                "Please run `huggingface-cli login` in your Colab notebook to update it."
+            )
+        else:
+            return base_message + (
+                "Please make sure you are properly logged in by executing `huggingface-cli login`, "
+                "or provide a valid token."
+            )
```

Comment:
This can be removed now that `_get_token_error_message` is not used anymore (all the logic is already contained in whoami)